lib/srv/desktop/rdp/rdpclient/src/client/global.rs

```diff
@@ -0,0 +1,129 @@
+/// If the handle is not found in the [`CLIENT_HANDLES`] map, it also returns [`CGOErrCode::ErrCodeFailure`].
+///
+/// Note that the return value does not represent the result of the function call itself,
+/// rather it represents whether or not request to call a function was successfully sent over
+/// a channel. Error handling for the function call itself is handled by [`Client`].
+pub fn call_function_on_handle(cgo_handle: CgoHandle, func: ClientFunction) -> CGOErrCode {
+    if let Some(handle) = CLIENT_HANDLES.get(cgo_handle) {
+        match handle.blocking_send(func) {
+            Ok(_) => return CGOErrCode::ErrCodeSuccess,
+            Err(_) => return CGOErrCode::ErrCodeFailure,
+        }
+    }
+
+    CGOErrCode::ErrCodeFailure
+}
+
+/// A global, static tokio runtime for use by all clients.
+#[dynamic]
+static TOKIO_RT: tokio::runtime::Runtime = tokio::runtime::Runtime::new().unwrap();
+
+/// A global, static map of [`ClientHandle`] indexed by [`CgoHandle`].
+///
+/// See [`ClientHandles`].
+#[dynamic]
+static CLIENT_HANDLES: ClientHandles = ClientHandles::new();
+
+const _: () = {
+    const fn assert_send_sync<T: Send + Sync>() {}
+    // Immutable references to following types can be used directly by multiple
+    // threads (goroutines) simultaneously, so we guarantee that they are Send + Sync.
+    assert_send_sync::<tokio::runtime::Runtime>();
+    assert_send_sync::<ClientHandles>();
+};
+
+/// A map of [`ClientHandle`] indexed by [`CgoHandle`].
+///
+/// A function can be dispatched to the [`Client`] corresponding to a
+/// given [`CgoHandle`] by retrieving it from this map and sending
+/// the desired function and arguments to the [`ClientFunctionDispatcher`].
+struct ClientHandles {
+    map: RwLock<HashMap<CgoHandle, ClientHandle>>,
+}
+
+impl ClientHandles {
+    fn new() -> Self {
+        ClientHandles {
+            map: RwLock::new(HashMap::new()),
+        }
+    }
+
+    fn insert(&self, cgo_handle: CgoHandle, dispatcher: ClientHandle) {
+        self.map.write().insert(cgo_handle, dispatcher);
+    }
+
+    pub fn get(&self, cgo_handle: CgoHandle) -> Option<ClientHandle> {
+        self.map.read().get(&cgo_handle).map(|c| (*c).clone())
+    }
+
+    fn remove(&self, cgo_handle: CgoHandle) {
+        self.map.write().remove(&cgo_handle);
+    }
+}
+
+/// `ClientHandle` is a type alias for a `Sender` that sends `ClientFunction` enums.
+///
+/// It is used to dispatch [`ClientFunction`]s to a corresponding [`FunctionReceiver`]
+/// on a `Client`.
+type ClientHandle = Sender<ClientFunction>;
+
+/// `FunctionReceiver` is a type alias for a `Receiver` that receives [`ClientFunction`]
+/// enums.
+///
+/// Each `Client` has a `FunctionReceiver` that it listens to for [`ClientFunction`]
+/// requests sent by its corresponding [`ClientHandle`], and subseqently calls a corresponding
+/// function whenever such requests are received.
+pub type FunctionReceiver = Receiver<ClientFunction>;
```

Comment:
The idea here is that what you're doing is like passing a `fn`. 

In reality you're passing a member of the `ClientFunction` enum. But since each of these corresponds 1-to-1 with a function on `Client`, and sending it causes the client to call that function, it's very much like you're calling a `fn` on that `Client`.

Reply:
How about I change `ClientFunction` to `FunctionRequest` and do

```rust
pub type FunctionRequestReceiver = Receiver<FunctionRequest>;
```

Then we can say we "send a `FunctionRequest` via the `ClientHandle` to the `FunctionRequestReceiver`, which executes the requested function."